### PR TITLE
Json schema validation

### PR DESCRIPTION
**Pull request by Fatme**
http://teampulse.telerik.com/view#item/280276

### CODE REVIEW

File: lib/json-schema/json-schema-loader.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+			var id = schema.id;
+			var extendsProperty = schema.extends;
+
+			if(!this.isSchemaLoaded(id)) {
+				if(extendsProperty && extendsProperty.length > 0) {
+					_.each(extendsProperty, (ext: ISchemaExtends) => {
+						var schemaRef = ext.$ref;
+						var extSchema = this.findSchema(schemaRef);
+
+						if(extSchema == null) {
```
Comment (rosen-vladimirov) [inline lib/json-schema/json-schema-loader.ts]:
```Typescript
if(extSchema) {
```
or 
```Typescript
if(extSchema === null) {
```

Comment (Fatme) [inline lib/json-schema/json-schema-loader.ts]: yep, should be `if(!extSchema)`

File: lib/json-schema/json-schema-loader.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+			_.each(fileNames, (fileName: string) => {
+				if( path.extname(fileName) === ".json") {
+					var fullFilePath = path.join(this.schemasFolderPath, fileName);
+					var schema = this.$fs.readJson(fullFilePath).wait();
+					this.schemas[schema.id] = schema;
+				}
+			});
+
+			var schemas = _.values(this.schemas);
+			_.each(schemas, (schema: ISchema) => this.loadSchema(schema).wait());
```
Comment (rosen-vladimirov) [inline lib/json-schema/json-schema-loader.ts]: Can we make these calls async? For example
```Typescript
var futs: IFuture[] =[];
_.each(schemas, (schema: ISchema) => {
     var f = this.loadSchema(schema);
     futs.push(f);
});

Future.wait(futs);
```
It just looks like there are some heavy operations here.

File: lib/json-schema/json-schema-resolver.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+				this.schema.extends.properties = Object.create(null);
+
+				this.buildValidationSchema(extendsProperty);
+			}
+		}
+
+		return this.schema;
+	}
+
+	private buildValidationSchema(extendsProperty: ISchemaExtends[]) {
```
Comment (rosen-vladimirov) [inline lib/json-schema/json-schema-resolver.ts]: missing return type (I'm not sure what is our convention, but I prefer to have them explicitly written)

File: lib/json-schema/json-schema-resolver.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+			if(refSchema && refSchema.properties) {
+				_.extend(this.schema.extends.properties, refSchema.properties);
+			}
+			if(refSchema.extends) {
+				this.buildValidationSchema(refSchema.extends);
+			}
+		});
+	}
+
+	private buildSchemaIdRegex(schemaId: string) {
```
Comment (rosen-vladimirov) [inline lib/json-schema/json-schema-resolver.ts]: again missing return type

File: lib/json-schema/json-schema-validator.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+			});
+		}
+
+		return this._validProperties;
+	}
+
+	public validate(data: IProjectData): any {
+		var validationErrors = this.getValidationErrors(data);
+		if(_.keys(validationErrors).length !== 0) {
+			var output = _.values(validationErrors).join("\n");
```
Comment (rosen-vladimirov) [inline lib/json-schema/json-schema-validator.ts]: maybe os.EOL instead of \n ?

File: lib/project/web-site-project.ts [hunk 6]
```diff
@@ -73,7 +77,7 @@ export class MobileWebSiteProject extends frameworkProjectBaseLib.FrameworkProje
 	}
 
 	public adjustBuildProperties(buildProperties: any, projectData?: IProjectData): any {
```
Comment (rosen-vladimirov) [inline lib/project/web-site-project.ts]: return type is any, but we return future? Maybe return type should be IFuture<any> ?

File: lib/services/project-properties-service.ts [hunk 7]
```diff
@@ -52,113 +54,49 @@ export class ProjectPropertiesService implements IProjectPropertiesService {
...
 			}
+			propertyValue = _.union(propertyValue, newValue);
+		} else {
+			this.$errors.fail("Unknown property update mode '%s'", mode);
+		}
 
-			if (propData.onChanging) {
-				this.$injector.dynamicCall(propData.onChanging, [propertyValue]).wait();
-			}
+		/* if(property === "FrameworkVersion") {
```
Comment (rosen-vladimirov) [inline lib/services/project-properties-service.ts]: commented code... Should we keep it?

File: lib/json-schema/json-schema-loader.ts (hunk 2 above)
Comment (teobugslayer) [inline lib/json-schema/json-schema-loader.ts]: We can combine the values() & each() calls

File: lib/json-schema/json-schema-loader.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	private loadSchemas(): IFuture<void> {
+		return (() => {
+			var fileNames = this.$fs.readDirectory(this.schemasFolderPath).wait();
+			_.each(fileNames, (fileName: string) => {
+				if( path.extname(fileName) === ".json") {
+					var fullFilePath = path.join(this.schemasFolderPath, fileName);
+					var schema = this.$fs.readJson(fullFilePath).wait();
+					this.schemas[schema.id] = schema;
+				}
+			});
```
Comment (teobugslayer) [inline lib/json-schema/json-schema-loader.ts]: Can we use [transform](https://lodash.com/docs#transform) here?

File: lib/json-schema/json-schema-validator.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+			this.$errors.fail(output);
+		}
+	}
```
Comment (teobugslayer) [inline lib/json-schema/json-schema-validator.ts]: return value is 'any', but there is no 'return' statement. Is this intended?

File: lib/json-schema/json-schema-validator.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+		}
+		return this._availableProperties;
+	}
+
+	private tryResolveValidationSchema(frameworkVersion: string, framework: string): ISchema {
+		var validationSchemaName = this.getValidationSchemaName(frameworkVersion, framework);
+		var schema = this.$jsonSchemaResolver.getSchema(validationSchemaName);
+
+		if(!schema) {
+			this.$errors.fail("Unable to resolve validation schema.");
```
Comment (teobugslayer) [inline lib/json-schema/json-schema-validator.ts]: Is there anything the user can do to fix the issue? If so, we can add it to the message.

**Comment by rosen-vladimirov** [comment]
Ok, the discussed issues (writing them here just to make sure we will not forget them):
1) require for "jsv" module should be require("JSV") as Linux is a case-sensitive guy (or girl, not sure).
2) for newly created project:
$ appbuilder prop print Framework
Unable to resolve framework 3.5.0Project. Valid frameworks are: Cordova, NativeScript or MobileWebSite
unable to resolve projectData

Also you have to rebase as currently I had to modify package.json to be able to use the code ;)

File: lib/json-schema/json-schema-validator.ts (hunk 9 above)
Comment (Fatme) [inline lib/json-schema/json-schema-validator.ts]: The return type should be `void` and I have already fixed it

**Comment by Fatme** [comment]
I fixed the issues mentioned from @rosen-vladimirov and rebased the branch on master.

File: lib/json-schema/json-schema-loader.ts (hunk 8 above)
Comment (Fatme) [inline lib/json-schema/json-schema-loader.ts]: It seems that we can, but I can't see what will be the benefit to use `transform`

**Comment by teobugslayer** [comment]
Good work, :+1:  after a green build

File: lib/json-schema/json-schema-validator.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+			});
+		}
+
+		return this._validPropertiesCache[key];
+	}
+
+	public validate(data: IProjectData): void {
+		var validationErrors = this.getValidationErrors(data);
+		if(_.keys(validationErrors).length !== 0) {
+			var output = _.values(validationErrors).join("\n");
```
Comment (rosen-vladimirov) [inline lib/json-schema/json-schema-validator.ts]: os.EOL maybe

File: lib/json-schema/json-schema-validator.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+			this.$errors.fail("Schema validation failed with following errors: \n %s", output);
```
Comment (rosen-vladimirov) [inline lib/json-schema/json-schema-validator.ts]: os.EOL maybe

File: lib/project/cordova-project.ts [hunk 13]
```diff
@@ -55,6 +57,14 @@ export class CordovaProject extends frameworkProjectBaseLib.FrameworkProjectBase
 		return _.values(allConfigFiles);
 	}
 
+	public getValidationSchemaName(frameworkVersion: string): string {
+		if(frameworkVersion && frameworkVersion.startsWith("2")) {
```
Comment (rosen-vladimirov) [inline lib/project/cordova-project.ts]: is it possible? I thought we are supporting version 3.0.0 and above

File: lib/services/project-properties-service.ts [hunk 14]
```diff
@@ -218,14 +179,14 @@ export class ProjectPropertiesService implements IProjectPropertiesService {
...
 			var propertyGroup: any = result.Project.PropertyGroup[0];
 
-			var projectSchema = frameworkProject.getFullProjectFileSchema().wait();
+			var projectSchema = frameworkProject.getProjectFileSchema();
 			_.sortBy(Object.keys(projectSchema), key => key === "FrameworkVersion" ? -1 : 1).forEach((propertyName) => {
 				if (propertyGroup.hasOwnProperty(propertyName)) {
 					properties[propertyName] = propertyGroup[propertyName][0];
 
-					if (projectSchema[propertyName].flags) {
+					/* if (projectSchema[propertyName].flags) {
```
Comment (rosen-vladimirov) [inline lib/services/project-properties-service.ts]: commented code

File: lib/services/project-properties-service.ts [hunk 15]
```diff
@@ -52,144 +53,104 @@ export class ProjectPropertiesService implements IProjectPropertiesService {
...
 	private getPropRange(propData: any): IFuture<string[]>{
 		return (() => {
 			if (propData.dynamicRange) {
 				return this.$injector.dynamicCall(propData.dynamicRange).wait();
 			}
+			if(propData.enum) {
+				return propData.enum;
+			}
+			if(propData.items) {
+				if(propData.items.enum) {
```
Comment (rosen-vladimirov) [inline lib/services/project-properties-service.ts]: maybe combine if and inner if:
```Typescript
if(propData.items && propData.items.enum) {
      return propData.items.enum;
}
```

**Comment by rosen-vladimirov** [comment]
:+1: 
 Great work! I have some minor comments, the only part that's concerning me is the commented code in lib/services/project-properties-service.ts  - is it there for future reference?

Anyway you have my approval, so "release the Kraken" :100:

**Pull request merged**